macOS-arm64 testbed tests failing

### Describe the bug

As of ~May 23, the macOS-arm64 testbed test has been reliably failing.


### How to reproduce the bug

Every CI run is currently affected.

### Minimum example code

```shell

```

### Screenshots

_No response_

### Environment details

- Operating system and version: macOS 15
- Python version: 3.12
- Software versions:
  - Toga: main @ 3050e07


### Logs

```
=========================== short test summary info ============================
FAILED tests/app/test_desktop.py::test_menu_close_windows - assert <toga.window.Window object at 0x1127aa240> not in <toga.window.WindowSet object at 0x104b5f440>
 +  where <toga.window.WindowSet object at 0x104b5f440> = <testbed.app.Testbed object at 0x101aa5be0>.windows
FAILED tests/app/test_desktop.py::test_menu_minimize - assert False
 +  where False = <tests_backend.window.WindowProbe object at 0x113034e60>.is_minimized
FAILED tests/app/test_document_app.py::test_save_document - AssertionError: expected call not found.
Expected: write(PosixPath('/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app/tests/app/docs/example.testbed'))
  Actual: not called.
FAILED tests/app/test_document_app.py::test_save_as_document - AssertionError: expected call not found.
Expected: write(PosixPath('/private/var/folders/p8/qyz0lmpd2mld64f_f4c66y4c0000gn/T/pytest-of-runner/pytest-0/test_save_as_document0/new_filename.testbed'))
  Actual: not called.
FAILED tests/widgets/canvas/test_canvas.py::test_press - AssertionError: Expected 'mock' to be called once. Called 0 times.
FAILED tests/widgets/canvas/test_canvas.py::test_activate - AssertionError: Expected 'mock' to be called once. Called 0 times.
FAILED tests/widgets/canvas/test_canvas.py::test_drag - AssertionError: Expected 'mock' to be called once. Called 0 times.
FAILED tests/widgets/test_detailedlist.py::test_select - AssertionError: assert None == <Row 13081b560 a='A1' b='B1' c=<toga.icons.Icon object at 0x1244af530> d='C1'>
 +  where None = <DetailedList:0x1308f9910>.selection
FAILED tests/widgets/test_detailedlist.py::test_deselect - assert None == <Row 130145a30 a=<tests.widgets.test_detailedlist.MyData object at 0x130145670> b=2 c=<toga.icons.Icon object at 0x1301458b0>>
 +  where None = <DetailedList:0x1307e09e0>.selection
FAILED tests/widgets/test_mapview.py::test_zoom - AssertionError: Zoom level 9: failed 0.352 < 3.0417242227122188 < 1.406
assert 3.0417242227122188 < 1.406
FAILED tests/widgets/test_multilinetextinput.py::test_undo_redo - AssertionError: assert 'Hello World!' == 'Hello'

  - Hello
  + Hello World!
FAILED tests/widgets/test_numberinput.py::test_undo_redo - AssertionError: assert Decimal('3.14159') == Decimal('3.14000')
 +  where Decimal('3.14159') = <NumberInput:0x130a46f60>.value
 +  and   Decimal('3.14000') = Decimal('3.14000')
FAILED tests/widgets/test_passwordinput.py::test_undo_redo - AssertionError: assert 'sekrit World!' == 'sekrit'

  - sekrit
  + sekrit World!
FAILED tests/widgets/test_table.py::test_select - AssertionError: assert None == <Row 173f270e0 a='A1' b='B1' c='C1' d='D1' e='E1'>
 +  where None = <Table:0x173e17aa0>.selection
FAILED tests/widgets/test_table.py::test_activate - AssertionError: assert None == <Row 1741a95b0 a='A1' b='B1' c='C1' d='D1' e='E1'>
 +  where None = <Table:0x1717f1c40>.selection
FAILED tests/widgets/test_table.py::test_multiselect - AssertionError: assert [] == [<Row 173bc1970 a='A1' b='B1' c='C1' d='D1' e='E1'>]

  Right contains one more item: <Row 173bc1970 a='A1' b='B1' c='C1' d='D1' e='E1'>

  Full diff:
  + []
  - [
  -     <Row 173bc1970 a='A1' b='B1' c='C1' d='D1' e='E1'>,
  - ]
FAILED tests/widgets/test_textinput.py::test_undo_redo - AssertionError: assert 'Hello World!' == 'Hello'

  - Hello
  + Hello World!
FAILED tests/widgets/test_tree.py::test_select - AssertionError: assert None == <Leaf Node 17452da30 a='A00' b='B00' c='C00' d='D00' e='E00'>
 +  where None = <Tree:0x17452c260>.selection
FAILED tests/widgets/test_tree.py::test_activate - AssertionError: assert None == <Leaf Node 173ef7da0 a='A00' b='B00' c='C00' d='D00' e='E00'>
 +  where None = <Tree:0x17416f5f0>.selection
FAILED tests/widgets/test_tree.py::test_multiselect - AssertionError: assert [] == [<Leaf Node 17454a7b0 a='A00' b='B00' c='C00' d='D00' e='E00'>]

  Right contains one more item: <Leaf Node 17454a7b0 a='A00' b='B00' c='C00' d='D00' e='E00'>

  Full diff:
  + []
  - [
  -     <Leaf Node 17454a7b0 a='A00' b='B00' c='C00' d='D00' e='E00'>,
  - ]
FAILED tests/widgets/test_tree.py::test_row_changes - AssertionError: assert None == <Leaf Node 174e13470 a='ANEW' b='BNEW' c='CY'>
 +  where None = <Tree:0x174504c50>.selection
FAILED tests/widgets/test_tree.py::test_headerless_row_changes - AssertionError: assert None == <Leaf Node 174ba7a70 a='ANEW' b='BNEW' c='CY'>
 +  where None = <Tree:0x1745248c0>.selection
FAILED tests/window/test_window.py::test_visibility[Window-second_window_kwargs0] - assert False
 +  where False = <tests_backend.window.WindowProbe object at 0x174e5deb0>.is_minimized
FAILED tests/window/test_window.py::test_window_state_change[MainWindow-second_window_kwargs0-WindowState.NORMAL-WindowState.MINIMIZED] - AssertionError: Expected 'mock' to be called once. Called 0 times.
= 24 failed, 649 passed, 5 skipped, 37 xfailed, 2 retried in 405.82s (0:06:45) =
=== Messages dropped during live streaming (use `log show` to see what they were)
```


### Additional context

This may be related to the rollout of a new macOS-15-arm64 Actions image. All the failing tests are using 20260520.0085.1; The previous jobs ran with 20260427.0018.1.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -353,7 +353,7 @@ jobs:
 
         - backend: "macOS-arm64"
           platform: "macOS"
-          runs-on: "macos-latest"
+          runs-on: "macos-14"
           app-user-data-path: "$HOME/Library/Application Support/org.beeware.toga.testbed"
 
         # We use a fixed Ubuntu version rather than `-latest` because at some point,

--- a/changes/4415.misc.md
+++ b/changes/4415.misc.md
@@ -1,0 +1,1 @@
+A workaround was added for a flaky Qt webview test.

--- a/changes/4427.bugfix.md
+++ b/changes/4427.bugfix.md
@@ -1,0 +1,1 @@
+Initial application focus on macOS is now more reliable on macOS 26.

--- a/cocoa/src/toga_cocoa/app.py
+++ b/cocoa/src/toga_cocoa/app.py
@@ -1,7 +1,9 @@
 import asyncio
+import platform
 import sys
 from pathlib import Path
 
+from packaging.version import Version
 from rubicon.objc import (
     SEL,
     NSMutableDictionary,
@@ -312,6 +314,11 @@ class App:
             self.native.setActivationPolicy(NSApplicationActivationPolicyAccessory)
         else:
             self.native.setActivationPolicy(NSApplicationActivationPolicyRegular)
+
+        # macOS 14 introduced a new API for giving focus to the app
+        # We can't run CI on earlier versions, so this is no-branch.
+        if Version(platform.mac_ver()[0]) >= Version("14.0"):  # pragma: no branch
+            self.native.activate()
 
     ######################################################################
     # App resources

--- a/cocoa/tests_backend/probe.py
+++ b/cocoa/tests_backend/probe.py
@@ -55,8 +55,8 @@ class BaseProbe:
         if toga.App.app.run_slow or wait_for:
             delay = max(1, delay)
 
+        print("Waiting for redraw" if message is None else message)
         if delay or wait_for:
-            print("Waiting for redraw" if message is None else message)
             if toga.App.app.run_slow or wait_for is None:
                 await asyncio.sleep(delay)
             else:

--- a/iOS/tests_backend/probe.py
+++ b/iOS/tests_backend/probe.py
@@ -12,8 +12,8 @@ class BaseProbe:
         if toga.App.app.run_slow or wait_for:
             delay = max(1, delay)
 
+        print("Waiting for redraw" if message is None else message)
         if delay or wait_for:
-            print("Waiting for redraw" if message is None else message)
             if toga.App.app.run_slow or wait_for is None:
                 await asyncio.sleep(delay)
             else:

--- a/qt/tests_backend/widgets/webview.py
+++ b/qt/tests_backend/widgets/webview.py
@@ -8,7 +8,13 @@ from .base import SimpleProbe
 
 class WebViewProbe(SimpleProbe):
     native_class = QWebEngineView
-    content_supports_url = True
+
+    # See #4415. Qt's Webview *does* support content URLs; however, there appears to be
+    # a bug with Qt's WebView that causes the URL to be returned as a `data:text/html`
+    # content URL rather than the requested URL. As a workaround for test reliability,
+    # disable tests for content URLs.
+    # content_supports_url = True
+    content_supports_url = False
     javascript_supports_exception = True
     supports_on_load = True
 

--- a/testbed/tests/widgets/test_webview.py
+++ b/testbed/tests/widgets/test_webview.py
@@ -62,7 +62,9 @@ async def assert_content_change(
         new_url = widget.url
         new_content = await get_content(widget)
 
-        changed = (new_url == url) and (new_content == content)
+        changed = (new_content == content) and (
+            not probe.content_supports_url or (new_url == url)
+        )
         if not changed:
             timer -= 0.05
             await asyncio.sleep(0.05)
@@ -210,7 +212,7 @@ async def test_static_content(widget, probe, on_load):
         widget,
         probe,
         message="Webview has static content",
-        url="https://example.com/" if probe.content_supports_url else None,
+        url="https://example.com/",
         content="<h1>Nice page</h1>",
         on_load=on_load,
         timeout=JS_TIMEOUT,
@@ -458,9 +460,10 @@ async def test_on_navigation_starting_sync(widget, probe, on_load):
         widget,
         probe,
         message="Webview has static content",
-        url="https://example.com/" if probe.content_supports_url else None,
+        url="https://example.com/",
         content="<h1>Nice page</h1>",
         on_load=on_load,
+        timeout=JS_TIMEOUT,
     )
 
     # test static content can be set with no URL
@@ -534,9 +537,10 @@ async def test_on_navigation_starting_async(widget, probe, on_load):
         widget,
         probe,
         message="Webview has static content",
-        url="https://example.com/" if probe.content_supports_url else None,
+        url="https://example.com/",
         content="<h1>Nice page</h1>",
         on_load=on_load,
+        timeout=JS_TIMEOUT,
     )
     # test url allowed by code
     await wait_for(

--- a/testbed/tests/widgets/test_webview.py
+++ b/testbed/tests/widgets/test_webview.py
@@ -35,7 +35,15 @@ async def get_content(widget):
     )
 
 
-async def assert_content_change(widget, probe, message, url, content, on_load):
+async def assert_content_change(
+    widget,
+    probe,
+    message,
+    url,
+    content,
+    on_load,
+    timeout=LOAD_TIMEOUT,
+):
     # Web views aren't instantaneous. Even for simple static changes of page
     # content, the DOM won't be immediately rendered. As a result, even though a
     # page loaded signal has been received, it doesn't mean the accessors for
@@ -45,7 +53,7 @@ async def assert_content_change(widget, probe, message, url, content, on_load):
     # *and* content to change in any way before asserting the new values.
 
     changed = False
-    timer = LOAD_TIMEOUT
+    timer = timeout
 
     await probe.redraw(message)
 
@@ -54,7 +62,7 @@ async def assert_content_change(widget, probe, message, url, content, on_load):
         new_url = widget.url
         new_content = await get_content(widget)
 
-        changed = new_url == url and new_content == content
+        changed = (new_url == url) and (new_content == content)
         if not changed:
             timer -= 0.05
             await asyncio.sleep(0.05)
@@ -150,6 +158,7 @@ async def test_clear_url(widget, probe, on_load):
         url=None,
         content="",
         on_load=on_load,
+        timeout=JS_TIMEOUT,
     )
 
 
@@ -168,6 +177,7 @@ async def test_load_empty_url(widget, probe, on_load):
         url=None,
         content="",
         on_load=on_load,
+        timeout=JS_TIMEOUT,
     )
 
 
@@ -203,6 +213,7 @@ async def test_static_content(widget, probe, on_load):
         url="https://example.com/" if probe.content_supports_url else None,
         content="<h1>Nice page</h1>",
         on_load=on_load,
+        timeout=JS_TIMEOUT,
     )
 
 
@@ -224,6 +235,7 @@ async def test_static_large_content(widget, probe, on_load):
         url=url,
         content=large_content,
         on_load=on_load,
+        timeout=JS_TIMEOUT,
     )
 
 

